### PR TITLE
Events

### PR DESCRIPTION
**Pull request by antkiewiczk**

### CODE REVIEW

**Comment by antkiewiczk** [comment]
Checked out https://github.com/wepublish/wepublish/pull/899 with squashed commits that doesn't show conflicts and commits don't have mixed authors.

File: libs/events/import/api/src/lib/import/events-import.service.ts [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+}
+
+interface ImportedEventsParams {
+  filter: ImportedEventFilter
+  order: 1 | -1
+  skip: number
+  take: number
+  sort: string
+  cacheManager: Cache
+  source: (typeof Providers)[keyof typeof Providers]
```
Comment (Itrulia) [inline libs/events/import/api/src/lib/import/events-import.service.ts:45]:
```suggestion
  source: Providers
```

File: libs/events/import/api/src/lib/import/events-import.service.ts [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+export class EventsImportService {
+  constructor(
+    @Inject(CACHE_MANAGER) private cacheManager: Cache,
+    private prisma: PrismaClient,
+    private mediaAdapter: MediaAdapterService
+  ) {}
+
+  async importedEvents({filter, order, skip, take, sort}: ImportedEventsResolverParams) {
+    const importableEvents = Promise.all(
+      providers.map(provider =>
```
Comment (Itrulia) [inline libs/events/import/api/src/lib/import/events-import.service.ts:219]: This should be injected and not hardcoded. For example like: https://github.com/nestjs/nest/issues/770#issuecomment-1412029060

File: libs/events/import/api/src/lib/import/events-import.service.ts [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+          cacheManager: this.cacheManager,
+          prisma: this.prisma,
+          mediaAdapter: this.mediaAdapter
+        })
+      }
+    }
+  }
+}
+
+class AgendaBasel implements EventsProvider {
```
Comment (Itrulia) [inline libs/events/import/api/src/lib/import/events-import.service.ts:265]: cachemanager & prisma should be passed down via Nestjs providers and not in methods.
Essentially make AgendaBasel an injectable service

File: libs/events/import/api/src/lib/import/events-import.service.ts [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+}
+
+export enum EventStatus {
+  Cancelled = 'CANCELLED',
+  Postponed = 'POSTPONED',
+  Rescheduled = 'RESCHEDULED',
+  Scheduled = 'SCHEDULED'
+}
+
+const parseAndCacheData = async (cacheManager: Cache, source: Providers): Promise<Event[]> => {
```
Comment (Itrulia) [inline libs/events/import/api/src/lib/import/events-import.service.ts:82]: Should probably be in a different file for AgendaBasel/Zürich

File: libs/events/import/api/src/lib/import/events-import.service.ts [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+    startsAt: start ? new Date(start.trim()) : null,
+    endsAt: end ? new Date(end.trim()) : null
+  }
+
+  return parsedEvent
+}
+
+const fetchAndTransformImage = async (url: string): Promise<ArrayBufferUpload> => {
+  try {
+    const response = await axios.get(url, {responseType: 'arraybuffer'})
```
Comment (Itrulia) [inline libs/events/import/api/src/lib/import/events-import.service.ts:190]: Can we use the HTTP service and not Axios directly?

File: libs/events/import/api/src/lib/import/events-import.model.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
+import {Field, ObjectType, InputType, registerEnumType, Int} from '@nestjs/graphql'
+import GraphQLJSON from 'graphql-type-json'
+
+export enum Providers {
```
Comment (Itrulia) [inline libs/events/import/api/src/lib/import/events-import.model.ts:4]: The problem with this is that all providers have to come by WePublish and can't be implemented by operators. Which imo defeats the purpose of it.

File: libs/events/import/api/src/lib/import/events-import.resolver.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  Mutations
+ */
+  @Mutation(returns => String, {
+    name: 'createEvent',
+    description: `
+      Creates and event based on data from importable events list and an id and provider.
+      Also, uploads an image to WePublish Image library.
+    `
+  })
+  createEvent(@Args('filter') filter: CreateEventArgs) {
```
Comment (Itrulia) [inline libs/events/import/api/src/lib/import/events-import.resolver.ts:51]: dont think `createEvent` is the most sensible name here as it's `createEventFromSource`

File: libs/api/src/lib/graphql/query.public.ts [hunk 8]
```diff
@@ -566,6 +566,12 @@ export const GraphQLPublicQuery = new GraphQLObjectType<undefined, Context>({
       resolve: (root, {id}, {prisma: {event}}) => getEvent(id, event)
     },
 
+    importedEventsIds: {
```
Comment (Itrulia) [inline libs/api/src/lib/graphql/query.public.ts:569]: Could have also been in V2

File: apps/api-example/schema-v2.graphql [hunk 9]
```diff
@@ -46,6 +51,43 @@ A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date
 """
 scalar DateTime
 
+type Event {
+  createdAt: DateTime!
+  description: JSON!
```
Comment (Itrulia) [inline apps/api-example/schema-v2.graphql:56]: Should be `Node!` from slate.

File: libs/events/import/api/src/lib/import/events-import.service.ts (hunk 5 above)
Comment (antkiewiczk) [reply libs/events/import/api/src/lib/import/events-import.service.ts:190]: axios is easier and it's not like I added it to the project, it was already there

Comment (Itrulia) [reply libs/events/import/api/src/lib/import/events-import.service.ts:190]: Yeah but this is not the Nextjs way, now you will have multiple axios instances and interceptors won't work as expected. Axios is only used for the `notify-jira.js`

File: libs/events/import/api/src/lib/import/agenda-basel.service.ts [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+  const upcomingEvents = events.filter((event: XMLEventType) => upcomingOnly(event))
+
+  // parse their structure to ours
+  const importedEvents = upcomingEvents?.map((a: XMLEventType) => {
+    return parseXMLEventToWpEvent(a, source)
+  })
+
+  // save in cache
+  const ttl = 8 * 60 * 60 * 1000 // 8 hours
+  await cacheManager.set('parsedEvents', importedEvents, ttl)
```
Comment (Itrulia) [inline libs/events/import/api/src/lib/import/agenda-basel.service.ts:53]: Wouldn't it make more sense to let the service decide how to cache? Then we also don't need to pass the cacheManager down.

File: libs/events/import/api/src/lib/import/agenda-basel.service.ts [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+import {CACHE_MANAGER} from '@nestjs/cache-manager'
+import {EventStatus, PrismaClient} from '@prisma/client'
+
+const AGENDA_BASEL_URL = 'https://www.agendabasel.ch/xmlexport/kzexport-basel.xml'
+
+const getFallbackDesc = (source: string) => `<p>Event imported from ${source}</p>`
+
+const parseAndCacheData = async (cacheManager: Cache, source: string): Promise<Event[]> => {
+  const parser = new xml2js.Parser()
+  const urlToQuery = AGENDA_BASEL_URL
```
[the remark this answers is not in the document]
Comment (antkiewiczk) [reply libs/events/import/api/src/lib/import/agenda-basel.service.ts:26]: but in order to be able to reuse this, Agenda Zurich would need to have the exact same structure, naming etc?

File: apps/api-example/schema-v2.graphql (hunk 9 above)
Comment (antkiewiczk) [reply apps/api-example/schema-v2.graphql:56]: are you able to point me where should I change the code and how? I am actually not sure 🤔 I know that this `description: JSON!` is autogenerated but I am not entirely sure from where exactly, and what should be the corrected source value

File: libs/events/import/api/src/lib/import/events-import.service.ts (hunk 5 above)
Comment (antkiewiczk) [reply libs/events/import/api/src/lib/import/events-import.service.ts]: if I understand correctly, I need to install `@nestjs/axios` package and inject it in agenda basel constructor like this `private readonly httpService: HttpService,`

https://docs.nestjs.com/techniques/http-module

Also, how to make sure to get the image as `arraybuffer`?

File: libs/events/import/api/src/lib/import/agenda-basel.service.ts (hunk 10 above)
Comment (antkiewiczk) [reply libs/events/import/api/src/lib/import/agenda-basel.service.ts:53]: so where would you put this logic?

Comment (Itrulia) [reply libs/events/import/api/src/lib/import/agenda-basel.service.ts:53]: in the service and pass it down

File: libs/events/import/api/src/lib/import/agenda-basel.service.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (Itrulia) [reply libs/events/import/api/src/lib/import/agenda-basel.service.ts:27]: Agenda Zurich is the same as Agenda Basel.

File: libs/events/import/api/src/lib/import/agenda-basel.service.ts
Comment (Itrulia) [inline libs/events/import/api/src/lib/import/agenda-basel.service.ts:1]: Just realised there's no tests for this one, could we add some?

File: libs/events/import/api/src/lib/import/agenda-basel.service.spec.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+  describe('AgendaBaselService', () => {
+    test('importedEvents method should return imported events from the cache if available', async () => {
+      const mockEventsDocument = {
+        nodes: [mockEvent],
+        pageInfo: {endCursor: '1', hasNextPage: false, hasPreviousPage: false, startCursor: '1'},
+        totalCount: 1
+      }
+      jest.spyOn(cacheManager, 'get').mockResolvedValueOnce([mockEvent])
+
+      jest.spyOn(service, 'importedEvents').mockResolvedValueOnce(mockEventsDocument)
```
Comment (Itrulia) [inline libs/events/import/api/src/lib/import/agenda-basel.service.spec.ts:97]: This test doesn't do anything as the thing that is being tried to test, is mocked

File: libs/events/import/api/src/lib/import/agenda-basel.service.spec.ts [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+        take: 10,
+        sort: ImportedEventSort.CREATED_AT
+      })
+      expect(result).toEqual(mockEventsDocument)
+    })
+
+    test('importedEvent should return the imported event by ID from the corresponding provider', async () => {
+      const id = '1'
+
+      jest.spyOn(service, 'importedEvent').mockResolvedValueOnce(mockEvent)
```
Comment (Itrulia) [inline libs/events/import/api/src/lib/import/agenda-basel.service.spec.ts:111]: This test doesn't do anything as the thing that is being tried to test, is mocked

File: libs/events/import/api/src/lib/import/agenda-basel.service.spec.ts [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+      const result = await service.importedEvent({id})
+
+      expect(service.importedEvent).toBeCalledWith({
```
Comment (Itrulia) [inline libs/events/import/api/src/lib/import/agenda-basel.service.spec.ts:115]: this asserts that the test is correct, not that the implementation is.

File: libs/ui/editor/src/lib/authorCheckPicker/author-check-picker.tsx
Comment (Itrulia) [inline libs/ui/editor/src/lib/authorCheckPicker/author-check-picker.tsx:1]: I think this one was the correct one, no?

File: libs/authentication/website/src/lib/session.context.ts [hunk 15]
```diff
@@ -1,10 +1,10 @@
 import {UserSession, User} from '@wepublish/website/api'
```
Comment (Itrulia) [inline libs/authentication/website/src/lib/session.context.ts:1]: Merge error here

File: libs/membership/api/src/lib/dashboard/subscriptions/dashboard-subscription.resolver.spec.ts [hunk 16]
```diff
@@ -149,7 +152,8 @@ describe('DashboardSubscriptionResolver', () => {
             where: {email: 'foo@wepublish.ch'},
             create: {
               active: true,
-              email: 'foo@wepublish.ch',
+              email: `${generateRandomString()}@wepublish.ch`,
```
Comment (Itrulia) [inline libs/membership/api/src/lib/dashboard/subscriptions/dashboard-subscription.resolver.spec.ts:155]: I guess these changes come from trying to fix things?

Comment (antkiewiczk) [reply libs/membership/api/src/lib/dashboard/subscriptions/dashboard-subscription.resolver.spec.ts:155]: yeah probably

File: libs/ui/editor/src/lib/authorCheckPicker/author-check-picker.tsx
Comment (antkiewiczk) [reply libs/ui/editor/src/lib/authorCheckPicker/author-check-picker.tsx:1]: I think now it's in panel as it used to and the imports are from the /panel/ directory as well

File: libs/membership/api/src/lib/dashboard/subscriptions/dashboard-subscription.resolver.spec.ts (hunk 16 above)
Comment (antkiewiczk) [reply libs/membership/api/src/lib/dashboard/subscriptions/dashboard-subscription.resolver.spec.ts]: not applicable anymore

**Review by Itrulia — APPROVED** [review]

**Review by tomaszdurka — APPROVED** [review]
lgtm - let's merge this one !

**Pull request merged**